filebeat/input/v2/statestore/resource.go

```diff
@@ -0,0 +1,313 @@
+	runtime.SetFinalizer(res, (*Resource).finalize)
+	return res
+}
+
+func (r *Resource) close() {
+	runtime.SetFinalizer(r, nil)
+	r.finalize()
+}
+
+func (r *Resource) finalize() {
+	defer r.unlink()
+
+	if r.IsLocked() {
+		r.doUnlock()
+	}
+}
+
+func (r *Resource) link(create bool) {
+	if r.entry != nil {
+		return
+	}
+
+	store := r.store
+	store.resourcesMux.Lock()
+	defer store.resourcesMux.Unlock()
+
+	entry := store.find(r.key)
+	if entry == nil && create {
+		entry = store.create(r.key)
+	}
+	r.entry = entry
+}
+
+// unlink removes the pointer to the memory backed resource.
+// If the in memory resource entry is not used by any other Resource,
+// then we will remove it from the table.
+func (r *Resource) unlink() {
+	if r.entry == nil {
+		return
+	}
+
+	entry := r.entry
+	r.entry = nil
+
+	store := r.store
+	store.resourcesMux.Lock()
+	defer store.resourcesMux.Unlock()
+	if entry.Release() {
+		store.remove(r.key)
+	}
+}
+
+// Lock locks a resource held by the store. It blocks until the lock becomes
+// available.
+func (r *Resource) Lock() {
+	checkNotLocked(r.IsLocked())
+
+	r.link(true)
+	r.entry.Lock()
+	r.isLocked = true
+}
+
+// TryLock attempts to lock the resource. It returns true if the lock has been
+// acquired successfully.
+func (r *Resource) TryLock() bool {
+	checkNotLocked(r.IsLocked())
+
+	r.link(true)
+	r.isLocked = r.entry.TryLock()
+	if !r.isLocked {
+		r.unlink()
+	}
+	return r.isLocked
+}
+
+// Unlock releases a resource.
+func (r *Resource) Unlock() {
+	checkLocked(r.IsLocked())
+
+	r.doUnlock()
+	r.close()
+}
+
+func (r *Resource) doUnlock() {
+	r.entry.Unlock()
+	r.markUnlocked()
+}
+
+// IsLocked checks if the resource currently holds the lock to the shared
+// registry entry.
+func (r *Resource) IsLocked() bool {
+	return r.isLocked
+}
+
+func (r *Resource) markLocked() {
+	r.isLocked = true
+}
+
+func (r *Resource) markUnlocked() {
+	r.isLocked = false
+}
+
+// Has check if resource is already in registry.
+// Has does not require the lock to be taken.
+func (r *Resource) Has() (bool, error) {
+	has := false
+	err := r.store.persistentStore.View(func(tx *registry.Tx) error {
+		found, err := tx.Has(registry.Key(r.key))
+		if err == nil {
+			has = found
+		}
+		return err
+	})
+
+	if err != nil {
+		err = &Error{
+			op:      "resource/has",
+			message: "failed to lookup resource",
+			cause:   err,
+		}
+	}
+	return has, err
+}
+
+// Update the registry state with fields in val.
+// If the resource key is unknown, then a new document is inserted into the
+// registry.
+// Update requires the resource to be locked.
+//
+// It is recommended to use Update only for resource meta-data updates,
+// that allow us to track and identify a resource. Read state updates
+// should be handled by UpdateOp.
+//
+// The update call is thread-safe, as the update operation itself is protected.
+// But data races are still possible, if any two go-routines update
+// an overlapping set of fields.
```

Comment:
I that case, I think in this case the implementation recommendation would be to fetch entry merge fields and update in the lock?

Reply:
A resource is just a 'key' for the store. But the key should be associated with some real-world resource like a file. Users wanting to collect a file must lock said file and only release the lock if the file should not be collected anymore. Like:

```
res, ok := store.TryLock(path)
if !ok {
  return <error>
}
defer res.Unlock()

// update file meta:
meta := struct{
  Path string
  LastOpen time.Time
  ...
}{ path, time.Now(), ... }

res.Update(&meta)

for { <tail file> }

```

Reply:
In the file collection use-case a resource might be updated by a file watcher and an active reader concurrently. This is perfectly valid, as the input holds the lock. In order to have a safe-update, the different go-routines must never update the same set of fields. The file watcher should update file metadata only (e.g. file is renamed), while the active reader should only update the offset. If go-routines adhere to updating distinct set of fields, there is no race.